File: Object-Detection-Using-HSV-Color-Space/detect-video.py
import cv2 as cv
import numpy as np

def noop(x): return None

#### Object detection from video using HSV color space
#### (HSV -> Hue, Saturation, Value)

## 1. Create "Set Mask" window with default HSV range to detect blue color
SET_MASK_WINDOW = "Set Mask"
cv.namedWindow(SET_MASK_WINDOW, cv.WINDOW_NORMAL)
cv.createTrackbar("Min Hue", SET_MASK_WINDOW, 90, 179, noop)
cv.createTrackbar("Max Hue", SET_MASK_WINDOW, 140, 179, noop)
cv.createTrackbar("Min Sat", SET_MASK_WINDOW, 74, 255, noop)
cv.createTrackbar("Max Sat", SET_MASK_WINDOW, 255, 255, noop)
cv.createTrackbar("Min Val", SET_MASK_WINDOW, 0, 255, noop)
cv.createTrackbar("Max Val", SET_MASK_WINDOW, 255, 255, noop)

## 2. Capture from default camera
videoCapture = cv.VideoCapture(0)

while True:
    ## 3. Capture video from camera and convert to HSV color space
    _, capturedVideo = videoCapture.read() # video
    # rotate video 180 degrees (it starts upside down on Windows)
    # capturedVideo = cv.flip(capturedVideo, 1) # Win                                
    capturedVideo = cv.resize(capturedVideo, None, None, fx=0.5, fy=0.5) # macOS
    capturedVideoHsv = cv.cvtColor(capturedVideo, cv.COLOR_BGR2HSV)

    ## 4. Get min and max HSV values from Set Mask window
    minHue = cv.getTrackbarPos("Min Hue", SET_MASK_WINDOW)
    maxHue = cv.getTrackbarPos("Max Hue", SET_MASK_WINDOW)
    minSat = cv.getTrackbarPos("Min Sat", SET_MASK_WINDOW)
    maxSat = cv.getTrackbarPos("Max Sat", SET_MASK_WINDOW)
    minVal = cv.getTrackbarPos("Min Val", SET_MASK_WINDOW)
    maxVal = cv.getTrackbarPos("Max Val", SET_MASK_WINDOW)
    minHsv = np.array([minHue, minSat, minVal])
    maxHsv = np.array([maxHue, maxSat, maxVal])

    ## 5. Create mask and result (masked) video
    # params: input array, lower boundary array, upper boundary array
    mask = cv.inRange(capturedVideoHsv, minHsv, maxHsv)
    # params: src1	array, src2 array, mask
    resultVideo = cv.bitwise_and(capturedVideo, capturedVideo, mask=mask)

    ## 6. Show videos
    cv.imshow("Captured Video", capturedVideo)
    # cv.imshow("Mask", mask)  // optional
    cv.imshow("Result (Masked) Video", resultVideo)
    if cv.waitKey(1) == 27: break   # Wait Esc key to end program

File: Object-Detection-Using-HSV-Color-Space/detect-video.cpp
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

/** Object detection from video using HSV color space
**  (HSV -> Hue, Saturation, Value)
**/

int main()
{
    //// 1. Create mask control window with default HSV range to detect blue color
    auto const MASK_WINDOW = "Mask Settings";
    cv::namedWindow(MASK_WINDOW, CV_WINDOW_AUTOSIZE);

    // HSV range to detect blue color
    int minHue = 90, maxHue = 140;
    int minSat = 74, maxSat = 255;
    int minVal =  0, maxVal = 255;

    // Create trackbars in mask settings window
    cvCreateTrackbar("Min Hue", MASK_WINDOW, &minHue, 179);
    cvCreateTrackbar("Max Hue", MASK_WINDOW, &maxHue, 179);
    cvCreateTrackbar("Min Sat", MASK_WINDOW, &minSat, 255);
    cvCreateTrackbar("Max Sat", MASK_WINDOW, &maxSat, 255);
    cvCreateTrackbar("Min Val", MASK_WINDOW, &minVal, 255);
    cvCreateTrackbar("Max Val", MASK_WINDOW, &maxVal, 255);

    //// 2. Capture from default camera
    cv::VideoCapture videoCapture(0);

    while (true) {
        //// 3. Capture and convert video to HSV color space
        cv::Mat inputVideo;
        videoCapture.read(inputVideo);
        cv::flip(inputVideo, inputVideo, 1);
        cv::Mat inputVideoHSV;
        cv::cvtColor(inputVideo, inputVideoHSV, cv::COLOR_BGR2HSV);

        //// 4. Create mask and result (masked) video
        cv::Mat mask;
        inRange(inputVideoHSV, cv::Scalar(minHue, minSat, minVal), cv::Scalar(maxHue, maxSat, maxVal), mask);
        cv::Mat resultVideo;
        bitwise_and(inputVideo, inputVideo, resultVideo, mask);

        //// 5. Show videos        
        imshow("Input Video", inputVideo);
        imshow("Result (Masked) Video", resultVideo);
        // imshow("Mask", mask);

        //// Wait for 'esc' (27) key press for 30ms. If pressed, end program.
        if (cv::waitKey(30) == 27) break;
    }
}